### PR TITLE
mds: add slow op tracker history size and threshold

### PR DESCRIPTION
**Pull request by fangwenjiang**
add mds slow op tracker history size (default 20) and threshold (default 20).

Signed-off-by: fangwenjiang <jiangfw@inspur.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/options/mds.yaml.in [hunk 1]
```diff
@@ -1105,6 +1105,18 @@ options:
...
+  type: uint
+  level: advanced
+  default: 20
+  with_legacy: true
+# track the op if over this threshold
+- name: mds_op_history_slow_op_threshold
+  type: float
+  level: advanced
+  default: 10
+  with_legacy: true
```
Comment (batrick) [inline src/common/options/mds.yaml.in:1119]: we don't add new options with `with_legacy`. Please rewrite without.

**Review by batrick — CHANGES_REQUESTED** [review]
please remove the `git` file (use `git reset -p HEAD^`)

File: src/common/options/mds.yaml.in (hunk 1 above)
Comment (fangwenjiang) [reply src/common/options/mds.yaml.in]: I have remove the "with_legacy" and git file, thanks!

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins retest this please

File: src/common/options/mds.yaml.in [hunk 2]
```diff
@@ -1105,6 +1105,16 @@ options:
   services:
   - mds
   with_legacy: true
+# Max number of slow ops to track
```
Comment (vshankar) [inline src/common/options/mds.yaml.in:1108]: recommend including desc/long_desc for the added options.

File: src/common/options/mds.yaml.in [hunk 3]
```diff
@@ -1105,6 +1105,16 @@ options:
...
+- name: mds_op_history_slow_op_size
+  type: uint
+  level: advanced
+  default: 20
+# track the op if over this threshold
+- name: mds_op_history_slow_op_threshold
+  type: float
+  level: advanced
+  default: 10
```
Comment (ajarr) [inline src/common/options/mds.yaml.in:1117]: mention, 
```
services:
- mds
```
?

File: src/common/options/mds.yaml.in [hunk 4]
```diff
@@ -1105,6 +1105,16 @@ options:
   services:
   - mds
   with_legacy: true
+# Max number of slow ops to track
+- name: mds_op_history_slow_op_size
+  type: uint
+  level: advanced
+  default: 20
+# track the op if over this threshold
+- name: mds_op_history_slow_op_threshold
```
Comment (ajarr) [inline src/common/options/mds.yaml.in:1114]: I guess description can be
desc: duration of a MDS op to be considered as a historical slow MDS op

**Review by ajarr — COMMENTED** [review]
In the commit message you say, history_slow_op_threshold default value is set to 20, but in the code you set the default value to 10. Modify the commit message?

File: src/mds/MDSRank.cc [hunk 5]
```diff
@@ -3766,6 +3769,9 @@ void MDSRankDispatcher::handle_conf_change(const ConfigProxy& conf, const std::s
   if (changed.count("mds_op_history_size") || changed.count("mds_op_history_duration")) {
     op_tracker.set_history_size_and_duration(conf->mds_op_history_size, conf->mds_op_history_duration);
   }
+  if (changed.count("mds_op_history_slow_op_size") || changed.count("mds_op_history_slow_op_threshold")) {
+    op_tracker.set_history_slow_op_size_and_threshold(conf->mds_op_history_slow_op_size, conf->mds_op_history_slow_op_threshold);
+  }
```
Comment (lxbsz) [inline src/mds/MDSRank.cc:3774]: Without the `with_legacy: true` could you access these two options directly ?

You can use `conf.get_val<>()` here instead.

**Pull request closed without merging**